### PR TITLE
Mesh reflector miniapp

### PR DESCRIPTION
**Pull request by dylan-copeland**
This PR adds a new miniapp for reflecting a high-order or NURBS hexahedral mesh about a plane. The output for now is a high-order mesh. For example,

./reflector -m ../../data/pipe-nurbs.mesh -n '0 0 1'
./reflector -m reflected.mesh -o '10 0 0' -n '1 0 0'
pipes

./reflector -m ../../data/fichera.mesh -o '1 0 0' -n '1 0 0'
./reflector -m reflected.mesh -o '0 0 1' -n '0 0 1'
fichera2

Also included is `HexMeshBuilder`, a class that may be more generally useful for constructing hexahedral meshes with valid orientations.

An unrelated change that is generally helpful is to make `GridFunction::GetVectorGradientHat` public.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3363](https://github.com/mfem/mfem/pull/3363) | @dylan-copeland | @pazner | @kmittal2 + @IdoAkkerman + @mlstowell | 12/16/22 | 2/15/23 | ⌛due 1/6/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: mesh/mesh.hpp [hunk 1]
```diff
@@ -223,6 +224,7 @@ class Mesh
    Array<int> be_to_edge;  // for 2D
    Table *bel_to_edge;     // for 3D
    Array<int> be_to_face;
+   mutable Table *face_to_elem;
```
Comment (mlstowell) [inline mesh/mesh.hpp:227]: This creates a bit of a quandary. This variable is used to cache the table created by `Mesh::GetFaceToElementTable`. However, the documentation for `Mesh::GetFaceToElementTable` clearly states that this table must be deleted by the caller implying that it is not a copy of internally cached data. Furthermore `Mesh::GetFaceToElementTable` recreates this table everything time it is called whereas `Mesh::FindFaceNeighbors` checks for the cached copy and uses that when present.

The new `face_to_elem` table is handled properly in this PR so that is not the issue. The question is should we treat `face_to_element` in the same manner as `face_edge` and `edge_vertex` i.e. cached internally and not deleted by the caller? This could break existing code unless we deprecate the existing `Mesh::GetFaceToElementTable` and create a new API to access the cached table. Another possibility is to keep the existing interface but alter `Mesh::GetFaceToElementTable` so that it copies the cached table rather than recreating it.

I don't know the best course of action but I think we should discuss this.

File: fem/gridfunc.hpp [hunk 2]
```diff
@@ -340,6 +338,8 @@ class GridFunction : public Vector
 
    void GetVectorGradient(ElementTransformation &tr, DenseMatrix &grad) const;
 
+   void GetVectorGradientHat(ElementTransformation &T, DenseMatrix &gh) const;
```
Comment (mlstowell) [inline fem/gridfunc.hpp:341]: If we make this public we should also document it (and perhaps GetVectorGradient so that people understand the distinction)

File: miniapps/meshing/reflector.cpp [hunk 3]
```diff
@@ -0,0 +1,1061 @@
...
+   add(p, diff, p);
+}
+
+class ReflectedCoefficient : public VectorCoefficient
+{
+private:
+   VectorCoefficient * a;
+   const Vector origin, normal;
+   Mesh *meshOrig;
+   Mesh *rmesh;
```
Comment (mlstowell) [inline miniapps/meshing/reflector.cpp:56]: Seems to be an unused variable.

File: miniapps/meshing/reflector.cpp [hunk 4]
```diff
@@ -0,0 +1,1061 @@
...
+         rbe->SetVertices(v);
+         reflected->AddBdrElement(rbe);
+
+         rbe = reflected->NewElement(orig_geom);
+         rbe->SetVertices(rv);
+         reflected->AddBdrElement(rbe);
+      }
+   }
+
+   for (int i=0; i<reflected->GetNBE(); ++i)
```
Comment (mlstowell) [inline miniapps/meshing/reflector.cpp:951]: When running the sample runs I saw a very familiar warning:
```Boundary elements with wrong orientation: 20 / 40 (fixed)```

Clearly we have an algorithm which can fix this so we could just make that part of the reflection process so that users don't get this warning. Ideally we could reorder the reflected boundary element vertices (similar to what is done for the reflected hexes) to avoid this.

Comment (dylan-copeland) [reply miniapps/meshing/reflector.cpp:951]: This is fixed now with 1 line, to swap the 0 and 2 vertices of each reflected quadrilateral boundary element.

File: miniapps/meshing/reflector.cpp (hunk 3 above)
Comment (dylan-copeland) [reply miniapps/meshing/reflector.cpp]: Removed.

File: mesh/mesh.hpp (hunk 1 above)
Comment (dylan-copeland) [reply mesh/mesh.hpp:227]: I see the issue as the inconsistency of `GetFaceToElementTable` and other functions like `GetEdgeVertexTable` and `GetFaceEdgeTable`. Only `GetFaceToElementTable` gives the caller ownership. Of course, this inconsistency existed before this PR, and as you mentioned, this PR handles `face_to_elem` properly.

I agree with your suggestion, that backwards compatibility can be preserved by copying the internal table `face_to_elem` in `GetFaceToElementTable` and returning a pointer to the copy. However, I don't see an advantage over the current implementation in this PR, which sets the internal `face_to_elem` table only if `FindFaceNeighbors` is called. Since `FindFaceNeighbors` is a new function, it is only used by the new miniapp. Of course, it could be used more in the future. My choice of the current implementation is based on the rare usage of `FindFaceNeighbors`, i.e. it makes a minimal change, only in the case that FindFaceNeighbors is used.

In summary, my view is that the current implementation is fine, and there isn't a serious problem here to address. It would be nice to make things more consistent, but backwards compatibility makes that challenging.

File: miniapps/meshing/makefile [hunk 5]
```diff
@@ -108,14 +108,14 @@ clean: clean-build clean-exec
 
 clean-build:
 	rm -f *.o *~ mobius-strip klein-bottle toroid twist
-	rm -f mesh-explorer shaper extruder trimmer
+	rm -f mesh-explorer shaper extruder trimmer reflector
```
Comment (mlstowell) [inline miniapps/meshing/makefile:111]: Should probably add `reflector` to `SEQ_MINIAPPS` above.

File: mesh/mesh.hpp (hunk 1 above)
Comment (mlstowell) [reply mesh/mesh.hpp:227]: I agree that this isn't a serious problem. I guess I'm mainly concerned with maintainability. For the benefit of future developers perhaps we could add some comments before the `face_to_element` variable in the header file explaining how it is used (or not) by `FindFaceNeighbors` and `GetFaceToElementTable`.

**Review by mlstowell — APPROVED** [review]
There are still a couple outstanding small comments but I'm sure they will be address. Thanks @dylan-copeland! This is a useful tool (and a challenge to other developers to add mixed element support).

**Comment by pazner** [comment]
@IdoAkkerman, @kmittal2, could you take a look at this PR when you get a chance? Thanks!

File: miniapps/meshing/makefile (hunk 5 above)
Comment (dylan-copeland) [reply miniapps/meshing/makefile:111]: Initially, I did add `reflector` to `SEQ_MINIAPPS`, but then I reverted that change in commit eadf460df55266d9b6f388ae8cab768e38416e8e, because it caused a CI failure. I added it back, and the CI is failing again. I could use some advice on how to fix this issue.

**Review by kmittal2 — COMMENTED** [review]
Had some discussion with Dylan offline and he will be addressing a couple minor comments I had..
It is not clear why the CI is failing so we still need to look into that before this PR can be merged.

**Review by kmittal2 — APPROVED** [review]

File: mesh/mesh.hpp (hunk 1 above)
Comment (dylan-copeland) [reply mesh/mesh.hpp]: I added some documentation explaining this situation for the 3 tables.

File: mesh/mesh.hpp [hunk 6]
```diff
@@ -1182,6 +1188,10 @@ class Mesh
    /// Return the indices and the orientations of all faces of element i.
    void GetElementFaces(int i, Array<int> &faces, Array<int> &ori) const;
 
+   /** @brief Returns the elements sharing a face with element @a elem,
+       including @a elem. */
+   void FindFaceNeighbors(const int elem, std::set<int> & nghb) const;
```
Comment (pazner) [inline mesh/mesh.hpp:1193]: Small nitpicks:

* The standard MFEM-style interface for this kind of function is to use `Array<int>` for the returned list of faces. You could use `Array::Unique` to deduplicate (or use a `std::set` in the implementation and convert to an `Array` at the end).
* I don't see a problem per-se with using `std::set` here, but I think consistency of the API has some value
* If we don't use `std::set`, we can remove `<set>` from the includes in this header.
* In cases like this where a type like `std::set` is used for an output parameter, I think it's better to just return the result by value: the intent is more clear from the function signature, it allows for better const-correctness, and it's basically free with copy elision/move semantics/RVO.

File: fem/gridfunc.hpp (hunk 2 above)
Comment (pazner) [reply fem/gridfunc.hpp:345]: These comments are a nice addition!

I think both should be marked `@brief`

File: mesh/mesh.hpp (hunk 6 above)
Comment (dylan-copeland) [reply mesh/mesh.hpp]: Good suggestion, and it is better to settle this now, to avoid API changes. I changed it to return an `Array`.

**Comment by pazner** [comment]
This should be ready once the CI failures are resolved. I think it may be an issue with the makefiles?

File: miniapps/meshing/makefile (hunk 5 above)
Comment (mlstowell) [reply miniapps/meshing/makefile:111]: I believe the issue is that `reflector` does not fit the expected pattern in that it produces a file named `reflected.mesh` rather than `reflector.mesh`. The script that produces the test runs assumes this pattern and since `reflector.mesh` doesn't exist it assumes the run failed. The `toroid` and `twist`miniapps also breaks this pattern.

Unfortunately, I don't think the script (located in `config/test.mk`) contains documentation explaining its assumptions but I think the following will work. Add a special `make` target for the `reflector` test which provides the name of the expected output file:
```
reflector-test-seq: %
	@$(call mfem-test-file,$<,, Meshing miniapp,reflected.mesh)
```

Comment (mlstowell) [reply miniapps/meshing/makefile:111]: There may still be other issues but this should help.

**Comment by dylan-copeland** [comment]
> This should be ready once the CI failures are resolved. I think it may be an issue with the makefiles?

@pazner the CI issue is fixed now, so I think this PR is ready.

File: miniapps/meshing/makefile (hunk 5 above)
Comment (dylan-copeland) [reply miniapps/meshing/makefile:113]: Thanks, a slight variation of this worked.

**Comment by pazner** [comment]
Thanks @dylan-copeland and @mlstowell!

**Comment by pazner** [comment]
Merged in `next` for testing...
